test/test-fs-readdir.c

```diff
@@ -0,0 +1,514 @@
+ * error codes.
+ */
+
+static int file_opendir_cb_count;
+
+static void file_opendir_cb(uv_fs_t* req) {
+  ASSERT(req == &opendir_req);
+  ASSERT(req->fs_type == UV_FS_OPENDIR);
+  ASSERT(req->result == UV_ENOTDIR);
+  ASSERT(req->ptr == NULL);
+
+  uv_fs_req_cleanup(req);
+  ++file_opendir_cb_count;
+}
+
+TEST_IMPL(fs_readdir_file) {
+  const char* path;
+  int r;
+
+  path = "test/fixtures/empty_file";
+  loop = uv_default_loop();
+
+  /* Fill the req to ensure that required fields are cleaned up */
+  memset(&opendir_req, 0xdb, sizeof(opendir_req));
+
+  /* Testing the synchronous flavor */
+  r = uv_fs_opendir(loop, &opendir_req, path, NULL);
+
+  ASSERT(r == UV_ENOTDIR);
+  ASSERT(opendir_req.fs_type == UV_FS_OPENDIR);
+  ASSERT(opendir_req.result == UV_ENOTDIR);
+  ASSERT(opendir_req.ptr == NULL);
+
+  uv_fs_req_cleanup(&opendir_req);
+
+  /* Fill the req to ensure that required fields are cleaned up */
+  memset(&opendir_req, 0xdb, sizeof(opendir_req));
+
+  /* Testing the async flavor */
+  r = uv_fs_opendir(loop, &opendir_req, path, file_opendir_cb);
+  ASSERT(r == 0);
+
+  ASSERT(file_opendir_cb_count == 0);
+
+  uv_run(loop, UV_RUN_DEFAULT);
+
+  ASSERT(file_opendir_cb_count == 1);
+
+  uv_fs_req_cleanup(&opendir_req);
+
+  MAKE_VALGRIND_HAPPY();
+  return 0;
+}
+
+/*
+ * This test makes sure that reading a non-empty directory with
+ * uv_fs_{open,read}_dir returns proper directory entries, including the
+ * correct entry types.
+ */
+
+static int non_empty_opendir_cb_count;
+static int non_empty_readdir_cb_count;
+static int non_empty_closedir_cb_count;
+
+static void non_empty_closedir_cb(uv_fs_t* req) {
+  ASSERT(req == &closedir_req);
+  ASSERT(req->result == 0);
+
+  ++non_empty_closedir_cb_count;
+}
+
+static void non_empty_readdir_cb(uv_fs_t* req) {
+  ASSERT(req == &readdir_req);
+  ASSERT(req->fs_type == UV_FS_READDIR);
+
+  // TODO jgilli: by default, uv_fs_readdir doesn't return UV_EOF when reading
+  // an emptry dir. Instead, it returns "." and ".." entries in sequence.
+  // Should this be fixed to mimic uv_fs_scandir's behavior?
+  if (req->result == 0) {
+    ASSERT(non_empty_readdir_cb_count == 5);
+    uv_fs_closedir(loop, &closedir_req, req->dir, non_empty_closedir_cb);
+  } else {
+    ASSERT(req->result == 1);
+    ASSERT(req->ptr == dirents);
+
+#ifdef HAVE_DIRENT_TYPES
+    if (!strcmp(dirents[0].name, "test_subdir") ||
+        !strcmp(dirents[0].name, ".") ||
+        !strcmp(dirents[0].name, "..")) {
+      ASSERT(dirents[0].type == UV_DIRENT_DIR);
+    } else {
+      ASSERT(dirents[0].type == UV_DIRENT_FILE);
+    }
+#else
+    ASSERT(dirents[0].type == UV_DIRENT_UNKNOWN);
+#endif /* HAVE_DIRENT_TYPES */
+
+    ++non_empty_readdir_cb_count;
+
+    uv_fs_readdir(uv_default_loop(),
+                  &readdir_req,
+                  req->dir,
+                  dirents,
+                  sizeof(dirents) / sizeof(dirents[0]),
+                  non_empty_readdir_cb);
+  }
+
+  uv_fs_req_cleanup(req);
```

Comment:
This isn't safe, is it?  `req == &readdir_req` and that's (re)initialized on line 375.